src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs

```diff
@@ -26,45 +25,109 @@ public SignatureTrustAndValidityVerificationProvider()
+                        status = SignatureVerificationStatus.Illegal;
+                    }
+                }
+                else if (primarySignatureVerificationSummary.Flags.HasFlag(SignatureVerificationStatusFlags.CertificateExpired))
+                {
+                    // We are not adding this log if the primary signature has a countersignature to check the expiration against the countersignature's timestamp.
+                    // If the countersignature shouldn't be check and the primary signature was expired, add this log.
+                    issues.Add(SignatureLog.Issue(!settings.AllowIllegal, NuGetLogCode.NU3011, string.Format(CultureInfo.CurrentCulture, Strings.SignatureNotTimeValid, nameof(AuthorPrimarySignature))));
+                }
+            }
+
+            return new SignedPackageVerificationResult(status, signature, issues);
+        }
+
+        private SignatureVerificationSummary VerifyValidityAndTrust(
+            Signature signature,
+            SignedPackageVerifierSettings timestampSettings,
+            SignatureVerifySettings settings,
+            X509Certificate2Collection certificateExtraStore,
+            List<SignatureLog> issues)
         {
             var timestampIssues = new List<SignatureLog>();
 
             Timestamp validTimestamp;
             try
             {
                 validTimestamp = signature.GetValidTimestamp(
-                    settings,
+                    timestampSettings,
                     _fingerprintAlgorithm,
                     timestampIssues);
             }
             catch (TimestampException)
             {
-                return new SignedPackageVerificationResult(SignatureVerificationStatus.Illegal, signature, timestampIssues);
-            }
+                issues.AddRange(timestampIssues);
 
-            var certificateExtraStore = signature.SignedCms.Certificates;
-            var signatureIssues = new List<SignatureLog>();
+                return new SignatureVerificationSummary(signature.Type, SignatureVerificationStatus.Illegal, SignatureVerificationStatusFlags.InvalidTimestamp);
+            }
 
             var status = signature.Verify(
                 validTimestamp,
                 settings,
                 _fingerprintAlgorithm,
                 certificateExtraStore,
-                signatureIssues);
+                issues);
 
-            signatureIssues.AddRange(timestampIssues);
+            issues.AddRange(timestampIssues);
 
-            return new SignedPackageVerificationResult(status, signature, signatureIssues);
+            return status;
         }
+
+        private bool ShouldFallbackToRepositoryCountersignature(SignatureVerificationSummary primarySignatureVerificationSummary)
+        {
+            return primarySignatureVerificationSummary.SignatureType == SignatureType.Author &&
```

Comment:
You could rename `primarySignatureVerificationSummary` to `primarySummary` to improve readability.